### PR TITLE
feat: ENR can now be updated

### LINKED ISSUE #1843: feat: update ENR based on RPC topic subscribe
### Problem
Subscribing to a new topics via RPC does not update the ENR when using shards.

### Suggested solution
discv5 has a way to update the record. Create the same in waku discv5.

### Alternatives considered
Restarting the node with new topics always works but is annoying.

### Additional context
https://github.com/status-im/nim-eth/blob/285da12bf318a2e21182dc2453a4a30c58f73067/eth/p2p/discoveryv5/protocol.nim#L247

### Acceptance criteria
- [ ] Subbing to a shard topic like ```/waku/2/rs/16/0``` via RPC should update the ENR in waku discv5.

### PR DESCRIPTION
**Pull request by SionoiS**
# Description

I added a way to update the Discv5 ENR and used a `AsyncEventQueue` to broadcast topic subscriptions.

# Changes
- [x] Discv5 add/remove shard from ENR
- [x] event system for topic subscriptions  

Fix #1843

Followed by #1918

### CODE REVIEW

File: apps/wakunode2/app.nim [hunk 1]
```diff
@@ -570,7 +570,7 @@ proc startRestServer(app: App, address: ValidIpAddress, port: Port, conf: WakuNo
   ## Relay REST API
   if conf.relay:
     let relayCache = TopicCache.init(capacity=conf.restRelayCacheCapacity)
-    installRelayApiHandlers(server.router, app.node, relayCache)
+    installRelayApiHandlers(server.router, app.node, app.wakuDiscv5, relayCache)
```
Comment (alrevuelta) [inline apps/wakunode2/app.nim:573]: It seems odd to pass here `wakuDiscv5`. Perhaps `wakuDiscv5` should be part of the `node`.

Not a blocker though.

Comment (SionoiS) [reply apps/wakunode2/app.nim:573]: It is weird but it's the simplest thing I could do.

A refactor of how handlers send info. internally would be good.

File: waku/waku_enr/sharding.nim [hunk 2]
```diff
@@ -68,7 +68,7 @@ func init*(T: type RelayShards, cluster: uint16, indices: seq[uint16]): T =
 
   RelayShards(cluster: cluster, indices: indicesSeq)
 
-func topicsToRelayShards*(topics: seq[string]): Result[Option[RelayShards], string] =
+proc topicsToRelayShards*(topics: seq[string]): Result[Option[RelayShards], string] =
```
Comment (jm-clius) [inline waku/waku_enr/sharding.nim:71]: Not major, but any specific reason for this change?

File: waku/waku_discv5.nim [hunk 3]
```diff
@@ -122,6 +122,75 @@ proc new*(T: type WakuDiscoveryV5,
 
   WakuDiscoveryV5.new(rng, conf, some(record))
 
+proc enrSharding(wd: WakuDiscoveryV5, newTopics: seq[PubsubTopic], add: bool):  Result[void, string] =
```
Comment (jm-clius) [inline waku/waku_discv5.nim:125]: Nitpick: perhaps a more verb-y proc name (`modifyEnrShardField` but something better 😛 )

**Review by jm-clius — APPROVED** [review]
Approving to not be a bottleneck, but I'm not sure I agree with passing in discv5 into the API. The API should have no knowledge about discovery in general. Wouldn't it be possible (and simpler) to just have a "watcher" task on `wakunode2/app.nim` that monitors the subscribed topics on the node and triggers the discv5 ENR update when it notices a new/removed topic?

File: waku/waku_enr/sharding.nim (hunk 2 above)
Comment (SionoiS) [reply waku/waku_enr/sharding.nim:71]: More in line with everything else. Also I'm still not sure I understand the diff? Is it no-side effect the only diff.

File: waku/waku_discv5.nim (hunk 3 above)
Comment (SionoiS) [reply waku/waku_discv5.nim:125]: Naming is hard...

**Comment by SionoiS** [comment]
> Approving to not be a bottleneck, but I'm not sure I agree with passing in discv5 into the API. The API should have no knowledge about discovery in general. Wouldn't it be possible (and simpler) to just have a "watcher" task on `wakunode2/app.nim` that monitors the subscribed topics on the node and triggers the discv5 ENR update when it notices a new/removed topic?

Yes it is a weird thing to do. I don't like your watcher idea either. Why watch for changes when the info could be sent directly.

I'll try to find something better. I wish I was more knowledgeable about NIm :(

**Comment by SionoiS** [comment]
I discussed with @richard-ramos go-waku way of sending the info. from the relay api to various waku sub-systems and it seams like more or less like I would have done it in Rust.

I still don't know what to do in Nim thought...

**Comment by SionoiS** [comment]
OK!

I'm now satisfied with this solution and we can expand upon it later too.

File: tests/wakunode_rest/test_rest_relay.nim [hunk 4]
```diff
@@ -95,7 +96,7 @@ suite "Waku v2 Rest API - Relay":
     topicCache.subscribe("pubsub-topic-3")
     topicCache.subscribe("pubsub-topic-x")
 
-    installRelayDeleteSubscriptionsV1Handler(restServer.router, node, topicCache)
+    installRelayDeleteSubscriptionsV1Handler(restServer.router, node,topicCache)
```
Comment (Ivansete-status) [inline tests/wakunode_rest/test_rest_relay.nim:99]: I think this file doesn't need to be changed

**Review by Ivansete-status — COMMENTED** [review]
Is this PR duplicated? I feel like a dejavu :)
( https://github.com/waku-org/nwaku/pull/1918 )

**Comment by SionoiS** [comment]
> Is this PR duplicated? I feel like a dejavu :) ( #1918 )

Yeah sorry should not have ask to review the follow-up PR before merging this.

**Pull request merged**